simmechanics/CMakeLists.txt

```diff
@@ -234,6 +234,34 @@ macro(generate_icub_simmechanics)
   neck_roll
 ")
     endif()
+    
+    # The ICUB_2_7 option is used to modify the iCub 2.5.5 model to 
+    # add some details of the iCub 2.7 until a proper iCub 2.7 model 
+    # is produced
+    set(ADDITIONAL_XML_BLOBS "")
+    if(GIVTWO_ICUB_2_7)
+      set(ADDITIONAL_XML_BLOBS
+"  - |
+         <link name="root_link_imu_frame"/>
+
+  - |
+         <joint name="root_link_imu_frame_fixed_joint" type="fixed">
+           <origin xyz="0.08515499999999998 -0.011000000000000022 -0.11230899999999999" rpy="-2.0943952105869323 0 -1.5707963267948963"/>
+           <parent link="root_link"/>
+           <child link="root_link_imu_frame"/>
+         </joint>
+  - |
+         <gazebo reference=\"root_link\">
```

Comment:
This should be `base_link`.

Please see the fix in  https://github.com/prashanthr05/icub-models/commit/d7e09b49930cb2e5b5ba036b7416875c8172363c

and a detailed comment inhttps://github.com/dic-iit/element_floating-base-estimation/issues/61#issuecomment-651695040

Posting the comment here if someone cannot access the private repository,

>We need to understand how the Gazebo IMU considers the sensor transform to generate the measurements. Given the CAD frame associated with the root IMU, the measurements from the Gazebo should reflect the gravity only along the zy-plane, more inclined towards the y-axis.

>Related to this,
In simulation, it was noticed that the measurements were not being published in the sensor local frame and it was always published with respect to the inertial frame. This was because the root_link_imu_acc sensor was referenced to root_link which is a massless link attached to the actual link base_link (base_link is the link actually identified by Gazebo). So this had lead to the measurements/orientations being published with respect to the root_link frame which had almost the same orientation as the Gazebo world except for the heading.

>This bug was identified and fixed in,
prashanthr05/icub-models@d7e09b4

>Now the root_link_imu_acc publishes measurements (accelerometer and gyroscope) in its local frame.

>cc @MiladShafiee also faced similar issues in simulation. Seems like the gazebo measurements coming from the port /icubSim/xsens_inertial is wrong.
It streams orientation rpy =0.0, 0.0, 0.0 in the case of static home position, instead of streaming the rootlink_R_imu orientation.

>However, the orientation still seems to be a problem.
The documentation here https://osrf-distributions.s3.amazonaws.com/gazebo/api/dev/classgazebo_1_1sensors_1_1ImuSensor.html#a66db2fd83d1d832184d888adc367115b says,

>ignition::math::Quaterniond Orientation ( ) const
get orientation of the IMU relative to a reference pose Initially, the reference pose is the boot up pose of the IMU, but user can call either SetReferencePose to define current pose as the reference frame, or call SetWorldToReferencePose to define transform from world frame to reference frame.
returns the orientation quaternion of the IMU relative to the imu reference pose.

>This might be related to osrf/gazebo#1959

Reply:
Mhh, I am a bit confused here. `base_link` and `root_link` share the same orientation, so I don't understand why there should be a difference. I typically prefer to attach the sensor to `root_link` as the fixed link lumping works in a different way in iDynTree/URDF/Rviz, in which "root_link" is the real link, while "base_link" is the massless additional frame.

Reply:
However, note that the problem is already there for the iCubGazeboV2_5_plus in https://github.com/robotology/icub-model-generator/blob/6e6ac9f5ecad68771f33c73257a9b02040a3f2d8/simmechanics/CMakeLists.txt#L145 .

Reply:
>Mhh, I am a bit confused here. base_link and root_link share the same orientation, so I don't understand why there should be a difference. I typically prefer to attach the sensor to root_link as the fixed link lumping works in a different way in iDynTree/URDF/Rviz, in which "root_link" is the real link, while "base_link" is the massless additional frame.

Possibly because the sensor not considering the massless link as its parent, which makes the world become its parent?

Reply:
> However, note that the problem is already there for the iCubGazeboV2_5_plus in

Indeed, I think I copied the changes to iCubGazeboV2_5 on my fork from the iCubGazeboV2_5_plus.

Reply:
> 
> 
> > Mhh, I am a bit confused here. base_link and root_link share the same orientation, so I don't understand why there should be a difference. I typically prefer to attach the sensor to root_link as the fixed link lumping works in a different way in iDynTree/URDF/Rviz, in which "root_link" is the real link, while "base_link" is the massless additional frame.
> 
> Possibly because the sensor is not considering the massless link as its parent and the relative pose we specify is not used, thereby making the world become its parent?

I think there is a bug in the URDF to SDF conversion on this, see https://github.com/robotology/icub-model-generator/issues/140 .

Reply:
So this might explain the problem with orientation as well.

Reply:
> So this might explain the problem with orientation as well described in #139 (comment). Possibly not.

Why not? I think it is exactly this the problem.

Reply:
Because even after changing the sensor's parent link to a lumped mass link (base_link), I remember having the problem with orientation measurements from the sensor. So I was thinking if it was more related to the initialization of IMU pose?

Reply:
Sorry, I referred indeed to the problem that was fixed by switching to using `base_link` as the reference. That seems indeed another problem that maybe is related to https://github.com/osrf/gazebo/issues/1959 .